### DISCUSSION TITLE
Optimization recommendations

### DISCUSSION THREAD

**Discussion by LKyao** [mpdf/mpdf#1879] [General]
When the HTML contains a large amount or large base64 image data, it can cause mPDF to throw an error indicating that it exceeds the pcre.backtrack_limit limit. Is there any way to optimize this?

**Comment by finwe** [comment]
https://stackoverflow.com/questions/49116886/uncaught-mpdf-mpdfexception-the-html-code-size-is-larger-than-pcre-backtrack-li/49126984#49126984

**Reply by finwe** [reply]
I'm afraid there will be no way around the base64 data, though. I can recommend only using static image files.

**Reply by LKyao** [reply]
> 不过，恐怕没有办法绕过base64数据。我建议只使用静态图像文件。

Thank you very much! The base64 image data is too large, which makes the PDF conversion process extremely time-consuming... sigh. The problem is that our business logic involves populating data in a Word template using PHPWord to generate an HTML file. The Word template includes image descriptions, so the generated HTML file contains images in base64 format. However, when we attempt to convert this HTML file to PDF using mPDF, it throws an error indicating that it exceeds the pcre.backtrack_limit limit.

**Reply by LKyao** [reply]
> 不过，恐怕没有办法绕过base64数据。我建议只使用静态图像文件。

I have an immature idea, and I'm not sure if it's feasible or not. The idea is to check for the presence of base64 image data when converting the HTML data. If there is any, can we temporarily save those images and replace the base64 data? Then, after generating the PDF, we can proceed with deleting the temporary images.

**Reply by LKyao** [reply]
> I'm afraid there will be no way around the base64 data, though. I can recommend only using static image files.

The approach is feasible, and the problem has been resolved. Haha.

```
function base64ToImgHTML($html){
        $dom = new \DOMDocument();
        $dom->loadHTML($html);
        $images = $dom->getElementsByTagName('img');

        foreach ($images as $image) {
            $src = $image->getAttribute('src');

            if (strpos($src, 'data:image') === 0) {
                $data = explode(',', $src)[1];
                $imageData = base64_decode($data);

                $tempFilename = uniqid() . '.png';
                $tempFilePath = 'tem/' . $tempFilename;
                file_put_contents($tempFilePath, $imageData);

                $image->setAttribute('src', $tempFilePath);
            }
        }
        return $dom->saveHTML();
    }
```
